[RyuJIT/ARM32] Enable passing large split struct argument

hseok-oh opened the pull request:
This enables passing split struct larger than 16 bytes.
To support splitted struct, it defines new GenTree type - GenTreePutArgSplit.
GenTreePutArgSplit is similar with GenTreePutArgStk,
but it is used for splitted struct only
and it has additional field to save register information.
GenTreePutArgSplit node is generated in lower phase.

related issue: #10722 
cc/ @dotnet/arm32-contrib

BruceForstall commented:
cc @dotnet/jit-contrib

CarolEidt commented on src/jit/gentree.h:
```diff
@@ -4845,6 +4856,86 @@ struct GenTreePutArgStk : public GenTreeUnOp
+    }
+
+    // not consider HFA yet
+    regNumber gtOtherRegs[MAX_REG_ARG - 1];
```
This comment is not clear. Should it be "HFA args are not yet handled"?

CarolEidt commented on src/jit/gentree.h:
```diff
@@ -4845,6 +4856,86 @@ struct GenTreePutArgStk : public GenTreeUnOp
+    // Return Value:
+    //     Return regNumber of ith register of this struct argument
+    //
+    regNumber GetRegNumByIdx(unsigned idx) const
```
I would spell out "Index" in both the method name and the parameter name.

hseok-oh replied:
But GenTreeCall, GenTreeCopyOrReload also uses same name. Isn't it better to use same name?

CarolEidt replied:
IMO there are a number of existing names that aren't ideal, but I'm OK with being consistent or switching.

hseok-oh replied:
I'll make new PR to fix this.

CarolEidt commented on src/jit/lower.cpp:
```diff
@@ -808,6 +813,37 @@ GenTreePtr Lowering::NewPutArg(GenTreeCall* call, GenTreePtr arg, fgArgTabEntryP
 #endif // !FEATURE_UNIX_AMD64_STRUCT_PASSING
 
+#ifdef _TARGET_ARM_
+    // Strcut can be splitted into register(s) and stack in ARM
```
Typos: I suggest "Struct can be split between register(s) and stack on ARM".

CarolEidt commented on src/jit/lsra.cpp:
```diff
@@ -4070,6 +4075,15 @@ void LinearScan::buildRefPositionsForNode(GenTree*                  tree,
+        if (tree->OperIsPutArgStruct())
+        {
+            // get i-th candidate
+            currCandidates = genFindLowestReg(candidates);
```
This relies on the fact that the arg regs are used in order. It would be good to add an assert that this is the same register as returned by `GetRegNumByIdx` (or `GetRegNumByIndex` if you change it).

hseok-oh replied:
We cannot compare registers return by `GetRegNumByIdx` here because it is not initialized here (LSRA buildInterval phase). It is initialized at LSRA resolve phase.

CarolEidt replied:
Ah, I was confused. Instead, how about asserting that the register is `genRegArgNext` of the previous register, if any?

CarolEidt commented on src/jit/morph.cpp:
```diff
@@ -1571,6 +1586,8 @@ void fgArgInfo::ArgsComplete()
                 curArgTabEntry->needTmp = true;
             }
+#ifndef _TARGET_ARM_
+            // Not tested yet
```
This should be some form of `TODO` comment, and there should presumably be an associated `NYI`?

hseok-oh replied:
If we uses NYI, legacyJIT/arm32 will be not works. I think It's better to remain it as `TODO` comment.

hseok-oh replied:
~~If we uses NYI, legacyJIT/arm32 will be not works.~~
Sorry, It's my misunderstand. But It doesn't seem important point for correctness. So anyway, I think It's better to remain it as TODO comment.

hseok-oh commented on src/jit/gentree.h:
```diff
@@ -1289,6 +1289,13 @@ struct GenTree
     }
 
+#ifdef _TARGET_ARM_
+    bool OperIsPutArgStruct() const
```
Thanks. I'll fix it.

hseok-oh commented:
@CarolEidt I've update your feedback. PTAL
- Reduce the clutter of #ifdef 
- Rename `PutArgStruct` as `PutArgSplit`
- Fix comments

CarolEidt commented on src/jit/codegenarmarch.cpp:
```diff
@@ -880,6 +886,216 @@ void CodeGen::genPutArgReg(GenTreeOp* tree)
 
+#ifdef _TARGET_ARM_
+//---------------------------------------------------------------------
+// genPutArgStruct - generate code for a GT_PUTARG_STRUCT node
```
This needs to be change to reflect the new name.

CarolEidt commented on src/jit/codegenarmarch.cpp:
```diff
@@ -880,6 +886,216 @@ void CodeGen::genPutArgReg(GenTreeOp* tree)
+
+    unsigned gcPtrCount; // The count of GC pointers in the struct
+    int      structSize;
+    bool     isHfa;
```
You should be able to get these values from the node. In inherits `gtNumberReferenceSlots` and `getArgSize()`, and it would make sense to add an `isHfa` flag (probably to `GenTreePutArgStk` to avoid extra padding).

hseok-oh replied:
I've update to get these value from the node.

CarolEidt commented on src/jit/codegenarmarch.cpp:
```diff
@@ -880,6 +886,216 @@ void CodeGen::genPutArgReg(GenTreeOp* tree)
+
+    // This is the varNum for our load operations,
+    // only used when we have a multireg struct with a LclVar source
+    unsigned varNumInp = BAD_VAR_NUM;
```
I'm not sure what the p is intended to imply in `varNumInp` is not very meaningful. Perhaps `srcVarNum` or `varNumIn` would be better?

CarolEidt commented on src/jit/codegenarmarch.cpp:
```diff
@@ -880,6 +886,216 @@ void CodeGen::genPutArgReg(GenTreeOp* tree)
+
+    while (remainingSize > 0)
+    {
+        {
```
Why the extra block here?

hseok-oh replied:
I'll fix it. Thanks.

CarolEidt commented on src/jit/codegenarmarch.cpp:
```diff
@@ -880,6 +886,216 @@ void CodeGen::genPutArgReg(GenTreeOp* tree)
+            else
+            {
+                // Need to handle additional loadSize cases here
+                noway_assert(loadSize == 4);
```
Should these be an `NYI` instead?

hseok-oh replied:
I think this assertion is needless for ARM32. This assertion comes form the implementation of `CodeGen::genPutArgStk. I'll remove this and fix the assertion in `CodeGen::genPutArgStk` later.

CarolEidt commented on src/jit/codegenarmarch.cpp:
```diff
@@ -1557,7 +1773,21 @@ void CodeGen::genCallInstruction(GenTreeCall* call)
+                    {
+                        regNumber regDst = (regNumber)((unsigned)argReg + idx);
+                        regNumber regSrc = (regNumber)((unsigned)argNode->gtRegNum + idx);
+                        inst_RV_RV(ins_Move_Extend(argNode->TypeGet(), argNode->InReg()), regDst, regSrc);
```
`InReg()` is no longer used for the RyuJIT backend, in which the code generator should always know whether a node is in a register or not. I think you need `true` here.

hseok-oh replied:
You're right. I'll update it.

hseok-oh commented:
@CarolEidt I've applied your feedback. PTAL
- Rename `varNumInp` to `srcVarNum`
- Fix comment: `genPutArgStruct` to `genPutArgSplit`
- Update to not use `InReg()`
- Get values from the node for code generation: `gcPtr`, `gcPtrCount`, `structSize`, `isHfa`
- Remove useless code

hseok-oh commented:
@dotnet-bot test Ubuntu x64 Checked Build and Test please

hseok-oh commented:
@dotnet-bot test Ubuntu x64 Checked Build and Test please

BruceForstall commented on src/jit/gtlist.h:
```diff
@@ -295,6 +295,9 @@ GTNODE(PINVOKE_PROLOG   , GenTree            ,0,GTK_LEAF|GTK_NOVALUE)   // pinvo
 GTNODE(PUTARG_REG       , GenTreeOp          ,0,GTK_UNOP)               // operator that places outgoing arg in register
 GTNODE(PUTARG_STK       , GenTreePutArgStk   ,0,GTK_UNOP|GTK_NOVALUE)   // operator that places outgoing arg in stack
+#ifdef _TARGET_ARM_
+GTNODE(PUTARG_SPLIT     , GenTreePutArgSplit ,0,GTK_UNOP)               // operator that places outgoing arg in registers with stack (splitted struct in ARM32)
```
To be precise, shouldn't this be:
```
#if defined(_TARGET_ARM_) && !defined(LEGACY_BACKEND)
```
?

BruceForstall replied:
>splitted [](start = 134, length = 8)

Grammatical note: do not use the word "splitted" (not a correct English word). Instead, simply use "split".

hseok-oh replied:
I'll fix it.

hseok-oh replied:
I'll fix it.

BruceForstall commented on src/jit/gtstructs.h:
```diff
@@ -96,6 +96,9 @@ GTSTRUCT_1(PhiArg      , GT_PHI_ARG)
 GTSTRUCT_N(Indir       , GT_STOREIND, GT_IND, GT_NULLCHECK, GT_BLK, GT_STORE_BLK, GT_OBJ, GT_STORE_OBJ, GT_DYN_BLK, GT_STORE_DYN_BLK)
 GTSTRUCT_1(PutArgStk   , GT_PUTARG_STK)
+#ifdef _TARGET_ARM_
+GTSTRUCT_1(PutArgSplit , GT_PUTARG_SPLIT)
```
==>
```
#if !defined(LEGACY_BACKEND) && defined(_TARGET_ARM_)
```
?

hseok-oh replied:
I'll fix it.

BruceForstall commented on src/jit/gentree.h:
```diff
@@ -4873,6 +4882,9 @@ struct GenTreePutArgStk : public GenTreeUnOp
     BYTE*    gtGcPtrs;               // gcPointers
+#ifdef _TARGET_ARM_
+    bool gtIsHfa; // isHfa
+#endif
```
The comment here is not useful (being a restatement of the member name). Is there something more descriptive that would be useful?

BruceForstall replied:
Is this actually required? Couldn't you use `Compiler::IsHfa(GenTreePtr tree)` to determine if this node is an HFA?

CarolEidt replied:
The call to `IsHfa` involves a JIT/EE interface call, so I think it's reasonable to maintain a flag on the node.

hseok-oh replied:
I'll remove this comment.

hseok-oh replied:
I agree with @CarolEidt

BruceForstall commented on src/jit/gentree.h:
```diff
@@ -4887,6 +4899,94 @@ struct GenTreePutArgStk : public GenTreeUnOp
 };
 
+#ifdef _TARGET_ARM_
+// Represent the struct argument: value is splitted in register(s) and stack
```
=>
```
#if defined(_TARGET_ARM_) && !defined(LEGACY_BACKEND)
```

BruceForstall commented on src/jit/gentree.h:
```diff
@@ -4887,6 +4899,94 @@ struct GenTreePutArgStk : public GenTreeUnOp
+
+    // HFA args is not yet handled
+    regNumber gtOtherRegs[MAX_REG_ARG - 1];
+
```
Can you add a small comment here about what this is, and its relationship to gtRegNum?

BruceForstall replied:
You can also mention that this is similar to GetRegNumByIdx/SetRegNumByIdx in GenTreeCall

---
In reply to: [122343854](https://github.com/dotnet/coreclr/pull/12050#discussion_r122343854) [](ancestors = 122343854)

hseok-oh replied:
I'll add comment.

BruceForstall commented on src/jit/gentree.h:
```diff
@@ -4887,6 +4899,94 @@ struct GenTreePutArgStk : public GenTreeUnOp
+    }
+
+    // HFA args is not yet handled
+    regNumber gtOtherRegs[MAX_REG_ARG - 1];
```
>regNumber [](start = 4, length = 9)

This should probably be `regNumberSmall`

hseok-oh replied:
Thanks. I'll fix it.

CarolEidt commented on src/jit/codegenarmarch.cpp:
```diff
@@ -1621,7 +1769,21 @@ void CodeGen::genCallInstruction(GenTreeCall* call)
             genConsumeReg(argNode);
             if (argNode->gtRegNum != argReg)
```
I think that this line should be moved to the `else` clause below, and in the `isSplit` case, each register needs to be consumed.

hseok-oh replied:
Thanks. I'll move this clause and implement `genConsumeReg`.

CarolEidt commented on src/jit/lower.cpp:
```diff
@@ -809,6 +808,40 @@ GenTreePtr Lowering::NewPutArg(GenTreeCall* call, GenTreePtr arg, fgArgTabEntryP
 #endif // !FEATURE_UNIX_AMD64_STRUCT_PASSING
 
+#ifdef _TARGET_ARM_
+    // Strcut can be split into register(s) and stack on ARM
```
Typo: Struct

BruceForstall replied:
>Strcut [](start = 7, length = 6)

Typo: "Strcut" => "Struct"

hseok-oh replied:
I'll fix it.

CarolEidt commented on src/jit/morph.cpp:
```diff
@@ -1557,6 +1565,14 @@ void fgArgInfo::ArgsComplete()
+#if defined(_TARGET_ARM_) && !defined(LEGACY_BACKEND)
+                else if (prevArgTabEntry->isSplit)
+                {
+                    // TODO: We need to test it on LegacyJIT/ARM32.
```
These should be something like `TODO-Arm`

hseok-oh replied:
I'll remove this comment because I did not found any problem on LegacyJIT/ARM32.

BruceForstall commented on src/jit/codegenarmarch.cpp:
```diff
@@ -1621,7 +1769,21 @@ void CodeGen::genCallInstruction(GenTreeCall* call)
+                if (curArgTabEntry->isSplit)
+                {
+                    for (unsigned idx = 0; idx < curArgTabEntry->numRegs; idx++)
+                    {
```
Could you add an assert:
```
assert(curArgTabEntry->numRegs >= 1);
```
?

hseok-oh replied:
I'll add this assertion.

BruceForstall commented on src/jit/lower.cpp:
```diff
@@ -809,6 +808,40 @@ GenTreePtr Lowering::NewPutArg(GenTreeCall* call, GenTreePtr arg, fgArgTabEntryP
+        if (arg->OperGet() != GT_OBJ)
+        {
+            NYI_ARM("Lowering: Oper for struct argument is not GT_OBJ");
+        }
```
This is an NYI and not an assert?

hseok-oh replied:
This NYI is for FIELD_LIST struct. I'll be enabling that later.

BruceForstall commented on src/jit/lower.cpp:
```diff
@@ -809,6 +808,40 @@ GenTreePtr Lowering::NewPutArg(GenTreeCall* call, GenTreePtr arg, fgArgTabEntryP
+        return putArg;
+    }
+#endif
+
```
Couldn't the end of this be:
```
    else
#endif // _TARGET_ARM_
```

in which case, you wouldn't need to include this in your `if`:
```
        if (arg->gtFlags & GTF_LATE_ARG)
        {
            putArg->gtFlags |= GTF_LATE_ARG;
        }
        else
        {
            info->node = putArg;
        }

...

        JITDUMP("new node is : ");
        DISPNODE(putArg);
        JITDUMP("\n");

        return putArg;
    }
#endif
```
?

(it looks like `updateArgTable` is always `true` so you could delete it.)

hseok-oh replied:
I'll fix it.

BruceForstall commented on src/jit/morph.cpp:
```diff
@@ -1439,6 +1439,14 @@ void fgArgInfo::ArgsComplete()
         }
+#if defined(_TARGET_ARM_) && !defined(LEGACY_BACKEND)
+        // Not tested on LegacyJIT/ARM32 yet
+        else if (curArgTabEntry->isSplit)
```
This will never be used on arm32 `LEGACY_BACKEND` -- we shouldn't be touching that code.

hseok-oh replied:
I'll remove this comment

BruceForstall commented on src/jit/lsraarmarch.cpp:
```diff
@@ -718,6 +732,57 @@ void Lowering::TreeNodeInfoInitPutArgStk(GenTreePutArgStk* argNode, fgArgTabEntr
+//
+// Arguments:
+//    argNode - a GT_PUTARG_STRUCT node
+//
```
This comment needs to be updated (it uses "struct" instead of "split" several places.)

hseok-oh replied:
I'll fix it.

BruceForstall commented on src/jit/morph.cpp:
```diff
@@ -1557,6 +1565,14 @@ void fgArgInfo::ArgsComplete()
+                {
+                    // TODO: We need to test it on LegacyJIT/ARM32.
+                    // It have worked well on LegacyJIT/ARM32 in spite of we did not set this.
+                    prevArgTabEntry->needPlace = true;
```
Can this comment be removed? We shouldn't be touching `LEGACY_BACKEND`.

hseok-oh replied:
I'll remove this comment.

BruceForstall commented on src/jit/morph.cpp:
```diff
@@ -1579,6 +1595,8 @@ void fgArgInfo::ArgsComplete()
             }
+#ifndef _TARGET_ARM_
+            // TODO: This implementation is not tested on ARM32
             else
```
Instead of `#ifdef`ing this out, can you make it a `NYI_ARM("multireg struct arguments");` ?

hseok-oh replied:
As I commented (https://github.com/dotnet/coreclr/pull/12050#discussion_r120794799), I think it's better to remain it as TODO comment.

BruceForstall commented on src/jit/morph.cpp:
```diff
@@ -3415,11 +3434,7 @@ GenTreeCall* Compiler::fgMorphArgs(GenTreeCall* call)
-                    }
-                    else if (size > 1 && size <= 4)
+                    if (isHfaArg || size > 1)
                     {
```
Why does this need to change? Is this a bug in existing `LEGACY_BACKEND`? Or a new requirement for RyuJIT? (To not check `size <= 4`)

hseok-oh replied:
Finally, there is no change because `fgMorphMultiregStructArg` skip morphing when size > 4.

BruceForstall commented on src/jit/codegenarmarch.cpp:
```diff
@@ -885,6 +891,148 @@ void CodeGen::genPutArgReg(GenTreeOp* tree)
+//
+// Arguments
+//    tree - the GT_PUTARG_STRUCT node
+//
```
here, and above, `GT_PUTARG_STRUCT` => `GT_PUTARG_SPLIT`

hseok-oh replied:
I'll fix it

hseok-oh commented on src/jit/codegenarmarch.cpp:
```diff
@@ -885,6 +891,148 @@ void CodeGen::genPutArgReg(GenTreeOp* tree)
+    unsigned structOffset  = nextIndex * TARGET_POINTER_SIZE;
+    int      remainingSize = structSize - structOffset;
+
+    // remainingSize is always multiple of TARGET_POINTER_SIZE
```
I'll add a assertion.

hseok-oh commented on src/jit/codegenarmarch.cpp:
```diff
@@ -885,6 +891,148 @@ void CodeGen::genPutArgReg(GenTreeOp* tree)
+    structOffset = 0;
+    for (unsigned idx = 0; idx < treeNode->gtNumRegs; idx++)
+    {
+        baseReg        = treeNode->GetRegNumByIdx(idx);
```
Good idea. I'll change this variable name.

hseok-oh commented:
@CarolEidt @BruceForstall  I've applied your feedback. PTAL

- Fix some comments: 
`genPutArgSplit`, `GenTreePutArgStk`, `GenTreePutArgSplit`, `NuwPutArg`, `ArgComplete`
- Add assertion check in `genPutArgSplit`, `genCallInstruction`
- Rename variable: `baseReg`
- Change flag for `GenTreePutArgSplit`: `_TARGET_ARM` && `!LEGACY_BACKEND`
- Change type of `gtOtherRegs` in `GenTreePutArgSplit`
- Remove duplicated code: `NewPutArg`
- Implement spill flag & restore for `GenTreePutArgSplit`'s `genConsumeReg`

hseok-oh commented:
Example 
```
using System;
using System.Runtime.CompilerServices;

public struct IntStruct
{
    public int num1;
    public int num2;
    public int num3;
    public int num4;
    public int num5;
    public int num6;
}

public class BringUpTest
{
    const int Pass = 100;
    const int Fail = -1;

    [MethodImplAttribute(MethodImplOptions.NoInlining)]
    public static int StructAdd(int a, int b, IntStruct rp, int c, int d)
    {   
        return rp.num1 + rp.num2 + rp.num3 + c + d + a + b + rp.num4 + rp.num5 + rp.num6;
    }

    public static int Main()
    {   
        IntStruct a = new IntStruct();

        a.num1 = 11;
        a.num2 = 12;
        a.num3 = 13;
        a.num4 = 14;
        a.num5 = 15;
        a.num6 = 16;

        int y = StructAdd(1, 2, a, 3, 4);
        if (y != 91) return Fail;
        return Pass;
    }
}
```

Generated Code (init struct & put argument)
```
Added IP mapping: 0x0008 STACK_EMPTY (G_M4347_IG02,ins#6,ofs#16)
Generating: N011 (  5,  6) [000011] -------------                IL_OFFSET void   IL offset: 0x8 REG NA
Generating: N013 (  1,  1) [000008] -------------        t8 =    CNS_INT   int    11 REG r3 $82
IN0007:             movs    r3, 11
                                                              /--*  t8     int
Generating: N015 (  5,  6) [000010] UA-----------             *  STORE_LCL_FLD int    V00 loc0         ud:3->0[+0] Fseq[num1] NA REG NA
IN0008:             str     r3, [sp+0x18]       // [V00 loc0]
Added IP mapping: 0x0011 STACK_EMPTY (G_M4347_IG02,ins#8,ofs#20)
Generating: N017 (  5,  6) [000017] -------------                IL_OFFSET void   IL offset: 0x11 REG NA
Generating: N019 (  1,  1) [000014] -------------       t14 =    CNS_INT   int    12 REG r3 $83
IN0009:             movs    r3, 12
                                                              /--*  t14    int
Generating: N021 (  5,  6) [000016] UA-----------             *  STORE_LCL_FLD int    V00 loc0         ud:4->0[+4] Fseq[num2] NA REG NA
IN000a:             str     r3, [sp+0x1c]       // [V00 loc0+0x04]
Added IP mapping: 0x001A STACK_EMPTY (G_M4347_IG02,ins#10,ofs#24)
Generating: N023 (  5,  6) [000023] -------------                IL_OFFSET void   IL offset: 0x1a REG NA
Generating: N025 (  1,  1) [000020] -------------       t20 =    CNS_INT   int    13 REG r3 $84
IN000b:             movs    r3, 13
                                                              /--*  t20    int
Generating: N027 (  5,  6) [000022] UA-----------             *  STORE_LCL_FLD int    V00 loc0         ud:5->0[+8] Fseq[num3] NA REG NA
IN000c:             str     r3, [sp+0x20]       // [V00 loc0+0x08]
Added IP mapping: 0x0023 STACK_EMPTY (G_M4347_IG02,ins#12,ofs#28)
Generating: N029 (  5,  6) [000029] -------------                IL_OFFSET void   IL offset: 0x23 REG NA
Generating: N031 (  1,  1) [000026] -------------       t26 =    CNS_INT   int    14 REG r3 $85
IN000d:             movs    r3, 14
                                                              /--*  t26    int
Generating: N033 (  5,  6) [000028] UA-----------             *  STORE_LCL_FLD int    V00 loc0         ud:6->0[+12] Fseq[num4] NA REG NA
IN000e:             str     r3, [sp+0x24]       // [V00 loc0+0x0c]
Added IP mapping: 0x002C STACK_EMPTY (G_M4347_IG02,ins#14,ofs#32)
Generating: N035 (  5,  6) [000035] -------------                IL_OFFSET void   IL offset: 0x2c REG NA
Generating: N037 (  1,  1) [000032] -------------       t32 =    CNS_INT   int    15 REG r3 $86
IN000f:             movs    r3, 15
                                                              /--*  t32    int
Generating: N039 (  5,  6) [000034] UA-----------             *  STORE_LCL_FLD int    V00 loc0         ud:7->0[+16] Fseq[num5] NA REG NA
IN0010:             str     r3, [sp+0x28]       // [V00 loc0+0x10]
Added IP mapping: 0x0035 STACK_EMPTY (G_M4347_IG02,ins#16,ofs#36)
Generating: N041 (  5,  6) [000041] -------------                IL_OFFSET void   IL offset: 0x35 REG NA
Generating: N043 (  1,  1) [000038] -------------       t38 =    CNS_INT   int    16 REG r3 $87
IN0011:             movs    r3, 16
                                                              /--*  t38    int
Generating: N045 (  5,  6) [000040] UA-----------             *  STORE_LCL_FLD int    V00 loc0         ud:8->0[+20] Fseq[num6] NA REG NA
IN0012:             str     r3, [sp+0x2c]       // [V00 loc0+0x14]
Added IP mapping: 0x003E STACK_EMPTY (G_M4347_IG02,ins#18,ofs#40)
Generating: N047 ( 39, 30) [000058] -------------                IL_OFFSET void   IL offset: 0x3e REG NA
Generating: N049 (  1,  1) [000045] -------------       t45 =    CNS_INT   int    3 REG r3 $88
IN0013:             movs    r3, 3
                                                              /--*  t45    int
Generating: N047 ( 39, 30) [000058] -------------                IL_OFFSET void   IL offset: 0x3e REG NA
Generating: N049 (  1,  1) [000045] -------------       t45 =    CNS_INT   int    3 REG r3 $88
IN0013:             movs    r3, 3
                                                              /--*  t45    int
Generating: N051 (???,???) [000086] -------------             *  PUTARG_STK [+0x10] int    REG NA
IN0014:             str     r3, [sp+0x10]       // [V01 OutArgs+0x10]
Generating: N053 (  1,  1) [000046] -------------       t46 =    CNS_INT   int    4 REG r3 $89
IN0015:             movs    r3, 4
                                                              /--*  t46    int
Generating: N055 (???,???) [000087] -------------             *  PUTARG_STK [+0x14] int    REG NA
IN0016:             str     r3, [sp+0x14]       // [V01 OutArgs+0x14]
Generating: N057 (  3,  2) [000044] -c-----N-----       t44 =    LCL_VAR_ADDR byref  V00 loc0         u:9 NA (last use) REG NA
                                                              /--*  t44    byref
Generating: N059 (  9,  7) [000051] -c-----------       t51 = *  OBJ(24)   struct REG NA $102
                                                              /--*  t51    struct
Generating: N061 (???,???) [000088] -------------       t88 = *  PUTARG_SPLIT struct REG r2
IN0017:             ldr     r3, [sp+0x20]       // [V00 loc0+0x08]
IN0018:             str     r3, [sp]    // [V01 OutArgs]
IN0019:             ldr     r3, [sp+0x24]       // [V00 loc0+0x0c]
IN001a:             str     r3, [sp+0x04]       // [V01 OutArgs+0x04]
IN001b:             ldr     r3, [sp+0x28]       // [V00 loc0+0x10]
IN001c:             str     r3, [sp+0x08]       // [V01 OutArgs+0x08]
IN001d:             ldr     r3, [sp+0x2c]       // [V00 loc0+0x14]
IN001e:             str     r3, [sp+0x0c]       // [V01 OutArgs+0x0c]
IN001f:             ldr     r2, [sp+0x18]       // [V00 loc0]
IN0020:             ldr     r3, [sp+0x1c]       // [V00 loc0+0x04]
Generating: N063 (  1,  1) [000042] -------------       t42 =    CNS_INT   int    1 REG r0 $8a
IN0021:             movs    r0, 1
                                                              /--*  t42    int
Generating: N065 (???,???) [000089] -------------       t89 = *  PUTARG_REG int    REG r0
Generating: N067 (  1,  1) [000043] -------------       t43 =    CNS_INT   int    2 REG r1 $8b
IN0022:             movs    r1, 2
                                                              /--*  t43    int
Generating: N069 (???,???) [000090] -------------       t90 = *  PUTARG_REG int    REG r1
Generating: N071 (  3,  7) [000091] -------------       t91 =    CNS_INT(h) int    0x744F4930 ftn REG lr
IN0023:             movw    lr, 0x4930
IN0024:             movt    lr, 0x744f
                                                              /--*  t91    int
Generating: N073 (  6,  9) [000092] -------------       t92 = *  IND       int    REG lr
IN0025:             ldr     lr, [lr]
                                                              /--*  t88    struct arg2 r2,r3
                                                              +--*  t89    int    arg0 in r0
                                                              +--*  t90    int    arg1 in r1
                                                              +--*  t92    int    control expr
Generating: N075 ( 35, 26) [000047] --C-G--------       t47 = *  CALL      int    BringUpTest.StructAdd $202
Call: GCvars=00000000 {}, gcrefRegs=0000 {}, byrefRegs=0000 {}
IN0026:             blx     lr          // BringUpTest:StructAdd(int,int,struct,int,int):int

```

BruceForstall commented on src/jit/codegenarmarch.cpp:
```diff
@@ -1617,11 +1766,31 @@ void CodeGen::genCallInstruction(GenTreeCall* call)
+                if (argNode->gtRegNum != argReg)
+                {
+                    {
+                        inst_RV_RV(ins_Move_Extend(argNode->TypeGet(), true), argReg, argNode->gtRegNum);
```
Why are there two scopes here?

hseok-oh replied:
My mistake. I'll fix it

BruceForstall commented on src/jit/gentree.h:
```diff
@@ -4887,6 +4899,165 @@ struct GenTreePutArgStk : public GenTreeUnOp
+    // more of its result regs are in that state.  The spill flag of each of the
+    // return register is stored in the below array.
+    unsigned gtSpillFlags[MAX_REG_ARG];
+
```
This looks the same as the `FEATURE_MULTIREG_RET` code in GenTreeCall. Can you adopt the recent change I made there, converting gtSpillFlags from an array of 4 32-bit unsigned elements (for arm32), namely 16 bytes, to a single 8-bit `unsigned char`? The key is that only data you need to store is 2 bits, for GTF_SPILL and GTF_SPILLED.

hseok-oh replied:
I'll rebase it and adopt recent change.

BruceForstall commented on src/jit/gentree.h:
```diff
@@ -4887,6 +4899,165 @@ struct GenTreePutArgStk : public GenTreeUnOp
+    //    Returns GTF_* flags associated with.
+    unsigned GetRegSpillFlagByIdx(unsigned idx) const
+    {
+        assert(idx < MAX_RET_REG_COUNT);
```
>MAX_RET_REG_COUNT [](start = 21, length = 17)

This should be:
```
assert(idx < MAX_REG_ARG);
```

BruceForstall commented on src/jit/codegenarmarch.cpp:
```diff
@@ -1617,11 +1766,31 @@ void CodeGen::genCallInstruction(GenTreeCall* call)
+                assert(curArgTabEntry->numRegs >= 1);
+                for (unsigned idx = 0; idx < curArgTabEntry->numRegs; idx++)
+                {
+                    regNumber regDst = (regNumber)((unsigned)curArgTabEntry->regNum + idx);
```
nit: instead of `regDst`, I would use `argReg`, since that is used elsewhere in the function to convey the same concept.

BruceForstall commented on src/jit/gentree.cpp:
```diff
@@ -1742,6 +1748,24 @@ regMaskTP GenTree::gtGetRegMask() const
+            {
+                regNumber reg = splitArg->GetRegNumByIdx(i);
+                if (reg != REG_NA)
+                {
```
Can this reg ever be REG_NA for split args? Or can you instead use `assert(reg != REG_NA)`?

hseok-oh replied:
I'll change this to use assert.

BruceForstall commented on src/jit/gentree.cpp:
```diff
@@ -11833,6 +11860,41 @@ void Compiler::gtGetArgMsg(
+                    else
+                    {
+                        regNumber otherRegNum = (regNumber)((unsigned)(argReg) + curArgTabEntry->numRegs - 1);
+                        char      seperator   = (curArgTabEntry->numRegs == 2) ? ',' : '-';
```
>seperator [](start = 34, length = 9)

typo: seperator => separator

BruceForstall commented on src/jit/gentree.cpp:
```diff
@@ -11833,6 +11860,41 @@ void Compiler::gtGetArgMsg(
+                {
+                    regNumber curReg = (regNumber)((unsigned)(argReg) + listCount);
+                    if (((unsigned)(curReg)) <= MAX_ARG_REG_COUNT)
+                    {
```
It seems wrong to cast a `regNumber` to `unsigned` and then compare it against a count. Should this instead by `if (listCount <= MAX_ARG_REG_COUNT)`?

hseok-oh replied:
For example, If `a(int a, int b, struct c)` and `sizeof(struct c)` is 8 byte, `struct c` uses `r2` and `r3`. So we need to compare MAX_ARG_REG_COUNT with the last register, not `listcount`.

BruceForstall replied:
Here's an alternative:
```
unsigned regArgNum = genMapIntRegNumToRegArgNum(argReg) + listCount;
if (regArgNum < MAX_REG_ARG)
{
    regNumber curReg = genMapIntRegArgNumToRegNum(regArgNum);
    sprintf_s(bufp, bufLength, "arg%d m%d %s%c", argNum, listCount, compRegVarName(curReg), 0);
}
else
{
    unsigned stackSlot = regArgNum - MAX_REG_ARG;
    sprintf_s(bufp, bufLength, "arg%d m%d out+%s%c", argNum, listCount, stackSlot, 0);
}
```

Notes:
1. `MAX_REG_ARG` is the maximum number of register arguments (for arm, that's r0, r1, r2, r3); `MAX_ARG_REG_COUNT` is the maximum number of argument registers that can be occupied by a single call argument.
2. The code above doesn't require casting between indices or offsets and registers.
3. I think the code before had a bug, by using `<=` instead of `<` -- it would allow curReg of 4, which would be r4, but the max argument register is r3.


---
In reply to: [122913222](https://github.com/dotnet/coreclr/pull/12050#discussion_r122913222) [](ancestors = 122913222)

hseok-oh replied:
OK. I understood it. Thank you for kind comment. I'll fix it.

hqueue replied:
@hseok-oh If so, than how about useing `REG_ARG_LAST` instead of `MAX_ARG_REG_COUNT` ?
Becasue `MAX_ARG_REG_COUNT` means maximum register count for a single argument, not regsiter number.
 as stated in `src/jit/target.h` below.
```
  #define MAX_ARG_REG_COUNT             4  // Maximum registers used to pass a single argument in multiple registers. (max is 4 floats or doubles using an HFA)
```

hqueue replied:
@hseok-oh In addition, how about using `REG_ARG_LAST` instead of `MAX_ARG_REG_COUNT` ?
Becasue `MAX_ARG_REG_COUNT` means maximum register count for a single argument, not register number.
 as stated in `src/jit/target.h` below.
```
  #define MAX_ARG_REG_COUNT             4  // Maximum registers used to pass a single argument in multiple registers. (max is 4 floats or doubles using an HFA)
```

hseok-oh replied:
@hqueue I fixed it based on @BruceForstall 's comment to consider HFA.

CarolEidt commented on src/jit/gentree.h:
```diff
@@ -4887,6 +4899,165 @@ struct GenTreePutArgStk : public GenTreeUnOp
+        ClearOtherRegFlags();
+    }
+
+    // First reg of struct is aloways given by gtRegNum.
```
Typo: aloways => always

CarolEidt commented on src/jit/lower.cpp:
```diff
@@ -809,112 +808,166 @@ GenTreePtr Lowering::NewPutArg(GenTreeCall* call, GenTreePtr arg, fgArgTabEntryP
+
+        // Set GC Pointer info
+        BYTE*    gcLayout = new (comp, CMK_Codegen) BYTE[info->numSlots + info->numRegs];
+        unsigned numRefs  = comp->info.compCompHnd->getClassGClayout(arg->gtObj.gtClass, gcLayout);
```
If you have a `GT_OBJ` node here, you can get the gc info from the `GenTreeObj` node, rather than having to make another call across the JIT\EE interface. And perhaps you could simply construct the layout for the `GT_FIELD_LIST` case once that is handled.

hseok-oh replied:
Thanks. I'll fix it.

hseok-oh replied:
@CarolEidt It seems that gc info is not initialized in this case. I got assertion failure when I tried to get gc info from `GT_OBJ`. So we need to generate gc info here.

CarolEidt replied:
The gc info should have been initialized when the `GT_OBJ` is created, unless it is known to have zero gc refs. That is, `IsGCInfoInitialized()` should return `true`, and `gtGcPtrs` should be non-null if `gtGcPtrCount` is non-zero. If that is not the case, then it is a bug.

hseok-oh commented:
@CarolEidt @BruceForstall I've applied your feedback. PTAL.
- Rebase
- Update managing spillFlag
- Implement spill & restore
- Fix typos and rename variables
- Fix bug related to print gentree

CarolEidt commented on src/jit/codegenarmarch.cpp:
```diff
@@ -1617,11 +1766,29 @@ void CodeGen::genCallInstruction(GenTreeCall* call)
         else
         {
-            regNumber argReg = curArgTabEntry->regNum;
             genConsumeReg(argNode);
```
It seems that you are still consuming only the first register here. Perhaps you need a `genConsumePutArgSplit` to consume all the register operands?

hseok-oh replied:
I updated `genUnspillRegIfNeeded` (https://github.com/hseok-oh/coreclr/blob/b5d86a6cec83522e6dcb784076f9136d81547129/src/jit/codegenlinear.cpp#L1007-L1035) called by `genConsumeReg`. 
And I added `gtSpillFlags` field in `GenTreePutArgSplit`. (https://github.com/hseok-oh/coreclr/blob/b5d86a6cec83522e6dcb784076f9136d81547129/src/jit/gentree.h#L4969)
To handle this flag, I also updated `genProduceReg` in `codegenlinear.cpp`, `rsSpillTree` and `rsUnspillInPlace` in `regset.cpp`. Is there more function to update to consume all registers?

CarolEidt replied:
The problem is that there are multiple registers, so `genConsumeReg()` needs to be called for each. That should take care of calling `genUnspillRegIfNeeded()`. The problem is that if you only call `genConsumeReg()` for the first register, then the liveness won't be updated for the other registers.

hseok-oh replied:
We can call `CodeGen::genConsumeReg(GenTree* tree)` to consume register for each argument . On other architectures, if argument is multi-reg struct (or long), it will be morphed to `FieldList` and each field has its `PutArgReg` node, and we can call `genConsumeReg()` for each `PutArgReg` node. But we don't morph to `FieldList` all multi-reg struct arg on arm32. For example, some parts of very large size struct can be passed using registers and IMO it makes too complex to morph this struct. So I defined new node type `PutArgSplit`, and it has multi-reg targets with stack slot. I updated this feature similar to multi-reg return.

CarolEidt replied:
I'm not sure what you're saying here. This code is not correct as-is, because multiple registers are being used, but you are not calling `genConsumeReg` for each one.

hseok-oh replied:
I cannot understand why we need to update the liveness when `genConsumeReg()` is called from `genCallInstruction()`. I thought that we should update liveness in `genConsumeReg()` when the argument `tree` of `genConsumeReg()` represents local variable (In other words, `tree->OperIsLocal() == true`). 
When `genConsumeReg()` is called from `genCallInstruction()`, `tree` does not represent local variable (it always has `GT_PUTARG_REG` as oper in general, and `GT_PUTARG_SPLIT` for split struct on arm32). As a result, updating livenesss in `genConsumeReg()` is skipped. So  I modified only `genUnspillRegIfNeeded()`.
Please tell me what I misunderstand.

CarolEidt replied:
If it is the case that a `fgArgTabEntry` with `isSplit == true` will always have a `GT_PUTARG_SPLIT`, then you should assert it here. In that case, please add a comment to indicate that a `genConsumeReg` is not needed, and move the `genConsumeReg` above to the else clause below here. Otherwise it makes no sense that you have consumed just a single register for a node that has multiple registers.

---
In reply to: [123937346](https://github.com/dotnet/coreclr/pull/12050#discussion_r123937346) [](ancestors = 123937346)

CarolEidt replied:
If it is the case that a `fgArgTabEntry` with `isSplit == true` will always have a `GT_PUTARG_SPLIT`, then you should assert it here. In that case, please add a comment to indicate that a `genConsumeReg` is not needed, and move the `genConsumeReg` above to the else clause below here. Otherwise it makes no sense that you have consumed just a single register for a node that has multiple registers.

hseok-oh replied:
Now I understand that you wanted to define special `genConsumeReg` function for `PutArgSplit` such as `genConsumeAddress`, `genConsumeRegs`, or `genConsumePutStructArgStk`, right? I also it's good approach because `PutArgSplit` consume register(s), and it will be better to handle liveness in this special function if need.

CarolEidt replied:
Yes. I think you are correct in that, if we always expect a `GT_PUTARG_SPLIT`, then it is sufficient to ensure that any spill or unspill is handled as needed.  But I think it would be best to handle this in a consistent way - e.g. with something like a `genConsumePutArgSplit`.

hseok-oh replied:
I defined new function `genConsumeArgSplitStruct`. PTAL

CarolEidt commented on src/jit/regset.cpp:
```diff
@@ -1535,8 +1538,15 @@ void RegSet::rsSpillTree(regNumber reg, GenTreePtr tree, unsigned regIdx /* =0 *
+        splitArg = tree->AsPutArgSplit();
+        treeType = splitArg->GetRegType(regIdx);
+    }
     else
```
This line is now under `#ifdef _TARGET_ARM_`, so for the non-arm case what is supposed to be the else clause of the `if (tree->IsMultiRegCall())` is unconditionally executed, which could possibly explain the x86 test failure.

hseok-oh commented:
@CarolEidt I've update
- Fix bug in regset.cpp
- Add comments in morph.cpp's NYI_ARM
- Fix comments' typo in lsraarmarcp.cpp
PTAL

CarolEidt commented:
@BruceForstall - I think this is ready to merge.

hseok-oh commented:
@CarolEidt @BruceForstall Thank you for review!
